File: bindings/java/src/test/java/com/deftio/triepack/ConformanceTest.java
/*
 * ConformanceTest.java
 *
 * Cross-language conformance suite.
 *
 * Reads tests/conformance/cases.txt -- the corpus every implementation shares
 * -- and checks this binding against the C-generated fixtures: decoding a
 * fixture must produce the corpus values, and encoding the corpus values must
 * produce the fixture byte for byte.
 *
 * See tests/conformance/README.md.
 *
 * Copyright (c) 2026 M. A. Chatterjee
 * SPDX-License-Identifier: BSD-2-Clause
 */

package com.deftio.triepack;

import static org.junit.jupiter.api.Assertions.*;

import java.io.IOException;
import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

class ConformanceTest {

    // ── Corpus location ───────────────────────────────────────────────

    private static Path conformanceDir() {
        // Tests run from bindings/java; walk up to the repository root.
        Path here = Paths.get("").toAbsolutePath();
        for (Path p = here; p != null; p = p.getParent()) {
            Path candidate = p.resolve("tests").resolve("conformance");
            if (Files.isDirectory(candidate)) {
                return candidate;
            }
        }
        throw new IllegalStateException("cannot locate tests/conformance from " + here);
    }

    private static Path fixturePath(String name) {
        return conformanceDir().resolve("fixtures").resolve(name + ".trp");
    }

    // ── Corpus parsing ────────────────────────────────────────────────

    /** Decode a corpus token ("~" for empty, %XX escapes) into bytes. */
    private static byte[] tokenBytes(String tok) {
        if (tok.equals("~")) {
            return new byte[0];
        }
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        for (int i = 0; i < tok.length(); ) {
            if (tok.charAt(i) == '%') {
                out.write(Integer.parseInt(tok.substring(i + 1, i + 3), 16));
                i += 3;
            } else {
                out.write(tok.charAt(i));
                i += 1;
            }
        }
        return out.toByteArray();
    }

    private static String tokenString(String tok) {
        return new String(tokenBytes(tok), StandardCharsets.UTF_8);
    }

    private static byte[] hexBytes(String tok) {
        if (tok.equals("~")) {
            return new byte[0];
        }
        byte[] out = new byte[tok.length() / 2];
        for (int i = 0; i < out.length; i++) {
            out[i] = (byte) Integer.parseInt(tok.substring(i * 2, i * 2 + 2), 16);
        }
        return out;
    }

    static final class Entry {
        final String key;
        final String type;
        final String arg;

        Entry(String key, String type, String arg) {
            this.key = key;
            this.type = type;
            this.arg = arg;
        }
    }

    static final class Case {
        final String name;
        final boolean encodeExpected;
        final List<Entry> entries = new ArrayList<>();

        Case(String name, boolean encodeExpected) {
            this.name = name;
            this.encodeExpected = encodeExpected;
        }

        Map<String, TpValue> data() {
            Map<String, TpValue> m = new LinkedHashMap<>();
            for (Entry e : entries) {
                m.put(e.key, corpusValue(e.type, e.arg));
            }
            return m;
        }

        byte[] fixture() throws IOException {
            return Files.readAllBytes(fixturePath(name));
        }

        @Override
        public String toString() {
            return name;
        }
    }

    /** Build the Java value a corpus entry describes. */
    private static TpValue corpusValue(String type, String arg) {
        switch (type) {
            case "null": return TpValue.ofNull();
            case "bool": return TpValue.ofBool("1".equals(arg));
            case "int": return TpValue.ofInt(Long.parseLong(arg));
            // uint64 values above Long.MAX_VALUE arrive as the matching
            // negative bit pattern, which is how TpValue carries them.
            case "uint": return TpValue.ofUInt(Long.parseUnsignedLong(arg));
            case "f64": return TpValue.ofFloat64(
                Double.longBitsToDouble(Long.parseUnsignedLong(arg, 16)));
            case "f32": return TpValue.ofFloat64(
                Float.intBitsToFloat((int) Long.parseUnsignedLong(arg, 16)));
            case "str": return TpValue.ofString(tokenString(arg));
            case "blob": return TpValue.ofBlob(hexBytes(arg));
            default: throw new IllegalArgumentException("unknown corpus type: " + type);
        }
    }

    static List<Case> parseCases() throws IOException {
        List<Case> cases = new ArrayList<>();
        Case cur = null;
        for (String raw : Files.readAllLines(conformanceDir().resolve("cases.txt"),
                                             StandardCharsets.UTF_8)) {
            String line = raw.trim();
            if (line.isEmpty() || line.startsWith("#")) {
                continue;
            }
            String[] parts = line.split(" ");
            if (parts[0].equals("case")) {
                boolean encodeExpected = !Arrays.asList(parts).contains("encode=no");
                cur = new Case(parts[1], encodeExpected);
                cases.add(cur);
            } else if (parts[0].equals("key")) {
                cur.entries.add(new Entry(tokenString(parts[1]), parts[2],
                                          parts.length > 3 ? parts[3] : null));
            } else {
                throw new IllegalArgumentException("unknown corpus directive: " + parts[0]);
            }
        }
        return cases;
    }

    static Stream<Case> cases() throws IOException {
        return parseCases().stream();
    }

    static Stream<Case> encodableCases() throws IOException {
        // Cases marked encode=no describe values this binding cannot reproduce
        // exactly (float32, or doubles some languages read back as integers).
        return parseCases().stream().filter(c -> c.encodeExpected);
    }

    /**
     * Doubles compare by bit pattern so -0.0 is distinguished, with all NaNs
     * treated as equal.
     */
    private static boolean valuesMatch(TpValue got, TpValue want) {
        if (got == null) {
            return false;
        }
        if (want.getType() == TpValue.Type.FLOAT64 && got.getType() == TpValue.Type.FLOAT64) {
            double g = got.float64Value();
            double w = want.float64Value();
            if (Double.isNaN(g) && Double.isNaN(w)) {
                return true;
            }
            return Double.doubleToRawLongBits(g) == Double.doubleToRawLongBits(w);
        }
        return got.equals(want);
    }

    // ── The suite ─────────────────────────────────────────────────────

    @Test
    void testCorpusAndFixturesPresent() throws IOException {
        List<Case> cases = parseCases();
        assertFalse(cases.isEmpty(), "corpus is empty");
        for (Case c : cases) {
            assertTrue(Files.exists(fixturePath(c.name)), "case " + c.name + ": missing fixture");
        }
    }

    @ParameterizedTest(name = "{0}")
    @MethodSource("cases")
    void testDecodesCFixture(Case c) throws IOException {
        Map<String, TpValue> result = TriePack.decode(c.fixture());
        Map<String, TpValue> want = c.data();
        assertEquals(want.size(), result.size(), "case " + c.name + ": key count");
        for (Map.Entry<String, TpValue> e : want.entrySet()) {
            assertTrue(valuesMatch(result.get(e.getKey()), e.getValue()),
                       "case " + c.name + ": key " + e.getKey()
                       + ": got " + result.get(e.getKey()) + ", want " + e.getValue());
        }
    }

    @ParameterizedTest(name = "{0}")
    @MethodSource("encodableCases")
    void testEncodesLikeCReference(Case c) throws IOException {
        assertArrayEquals(c.fixture(), TriePack.encode(c.data()), "case " + c.name);
    }

    // ── Malformed inputs ──────────────────────────────────────────────

    static Stream<String> malformedFixtures() throws IOException {
        try (Stream<Path> files = Files.list(conformanceDir().resolve("malformed"))) {
            return files.map(p -> p.getFileName().toString())
                        .filter(n -> n.endsWith(".trp"))
                        .sorted()
                        .collect(java.util.stream.Collectors.toList())
                        .stream();
        }
    }

    @Test
    void testMalformedCorpusPresent() throws IOException {
        assertTrue(malformedFixtures().findAny().isPresent(), "malformed corpus is empty");
    }

    /**
     * Each file is a valid fixture with one field damaged. Damage inside the
     * data is re-sealed with a correct CRC, so the reader has to catch it
     * rather than being handed a checksum failure.
     */
    @ParameterizedTest(name = "{0}")
    @MethodSource("malformedFixtures")
    void testMalformedInputRejected(String name) throws IOException {
        byte[] buf = Files.readAllBytes(conformanceDir().resolve("malformed").resolve(name));
        assertThrows(RuntimeException.class, () -> TriePack.decode(buf), name);
    }
}

File: bindings/kotlin/src/test/kotlin/com/deftio/triepack/ConformanceTest.kt
/*
 * ConformanceTest.kt
 *
 * Cross-language conformance suite.
 *
 * Reads tests/conformance/cases.txt -- the corpus every implementation shares
 * -- and checks this binding against the C-generated fixtures: decoding a
 * fixture must produce the corpus values, and encoding the corpus values must
 * produce the fixture byte for byte.
 *
 * See tests/conformance/README.md.
 *
 * Copyright (c) 2026 M. A. Chatterjee
 * SPDX-License-Identifier: BSD-2-Clause
 */

package com.deftio.triepack

import java.io.File
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ConformanceTest {

    // ── Corpus location ───────────────────────────────────────────────

    private val conformanceDir: File by lazy {
        // Tests run from bindings/kotlin; walk up to the repository root.
        var dir: File? = File("").absoluteFile
        while (dir != null) {
            val candidate = File(File(dir, "tests"), "conformance")
            if (candidate.isDirectory) return@lazy candidate
            dir = dir.parentFile
        }
        error("cannot locate tests/conformance")
    }

    private fun fixtureFile(name: String) = File(File(conformanceDir, "fixtures"), "$name.trp")

    // ── Corpus parsing ────────────────────────────────────────────────

    /** Decode a corpus token ("~" for empty, %XX escapes) into bytes. */
    private fun tokenBytes(tok: String): ByteArray {
        if (tok == "~") return ByteArray(0)
        val out = ArrayList<Byte>(tok.length)
        var i = 0
        while (i < tok.length) {
            if (tok[i] == '%') {
                out.add(tok.substring(i + 1, i + 3).toInt(16).toByte())
                i += 3
            } else {
                out.add(tok[i].code.toByte())
                i += 1
            }
        }
        return out.toByteArray()
    }

    private fun tokenString(tok: String) = String(tokenBytes(tok), Charsets.UTF_8)

    private fun hexBytes(tok: String): ByteArray {
        if (tok == "~") return ByteArray(0)
        return ByteArray(tok.length / 2) { tok.substring(it * 2, it * 2 + 2).toInt(16).toByte() }
    }

    private data class Entry(val key: String, val type: String, val arg: String?)

    private data class Case(
        val name: String,
        val encodeExpected: Boolean,
        val entries: MutableList<Entry> = mutableListOf()
    )

    private fun parseCases(): List<Case> {
        val cases = mutableListOf<Case>()
        File(conformanceDir, "cases.txt").forEachLine { raw ->
            val line = raw.trim()
            if (line.isNotEmpty() && !line.startsWith("#")) {
                val parts = line.split(" ")
                when (parts[0]) {
                    "case" -> cases.add(Case(parts[1], !parts.contains("encode=no")))
                    "key" -> cases.last().entries.add(
                        Entry(tokenString(parts[1]), parts[2], parts.getOrNull(3))
                    )
                    else -> error("unknown corpus directive: ${parts[0]}")
                }
            }
        }
        return cases
    }

    /** Build the Kotlin value a corpus entry describes. */
    private fun corpusValue(type: String, arg: String?): TpValue? = when (type) {
        "null" -> null
        "bool" -> TpValue.Bool(arg == "1")
        "int" -> TpValue.Int(arg!!.toLong())
        // uint64 values above Long.MAX_VALUE arrive as the matching negative
        // bit pattern, which is how TpValue carries them.
        "uint" -> TpValue.UInt(java.lang.Long.parseUnsignedLong(arg!!))
        "f64" -> TpValue.Float64(
            java.lang.Double.longBitsToDouble(java.lang.Long.parseUnsignedLong(arg!!, 16))
        )
        "f32" -> TpValue.Float64(
            java.lang.Float.intBitsToFloat(java.lang.Long.parseUnsignedLong(arg!!, 16).toInt())
                .toDouble()
        )
        "str" -> TpValue.Str(tokenString(arg!!))
        "blob" -> TpValue.Blob(hexBytes(arg!!))
        else -> error("unknown corpus type: $type")
    }

    private fun caseData(c: Case): Map<String, TpValue?> =
        c.entries.associate { it.key to corpusValue(it.type, it.arg) }

    /**
     * Doubles compare by bit pattern so -0.0 is distinguished, with all NaNs
     * treated as equal.
     */
    private fun valuesMatch(got: TpValue?, want: TpValue?): Boolean {
        if (want is TpValue.Float64 && got is TpValue.Float64) {
            if (got.value.isNaN() && want.value.isNaN()) return true
            return got.value.toRawBits() == want.value.toRawBits()
        }
        return got == want
    }

    // ── The suite ─────────────────────────────────────────────────────

    @Test
    fun testCorpusAndFixturesPresent() {
        val cases = parseCases()
        assertTrue(cases.isNotEmpty(), "corpus is empty")
        for (c in cases) {
            assertTrue(fixtureFile(c.name).exists(), "case ${c.name}: missing fixture")
        }
    }

    @Test
    fun testDecodesCFixtures() {
        for (c in parseCases()) {
            val result = decode(fixtureFile(c.name).readBytes())
            val want = caseData(c)
            assertEquals(want.size, result.size, "case ${c.name}: key count")
            for ((k, v) in want) {
                assertTrue(
                    valuesMatch(result[k], v),
                    "case ${c.name}: key $k: got ${result[k]}, want $v"
                )
            }
        }
    }

    /**
     * Each file is a valid fixture with one field damaged. Damage inside the
     * data is re-sealed with a correct CRC, so the reader has to catch it
     * rather than being handed a checksum failure.
     */
    @Test
    fun testRejectsMalformed() {
        val dir = File(conformanceDir, "malformed")
        val files = dir.listFiles { f: File -> f.name.endsWith(".trp") }?.sortedBy { it.name }
            ?: emptyList()
        assertTrue(files.isNotEmpty(), "malformed corpus is empty")
        for (f in files) {
            val bytes = f.readBytes()
            var rejected = false
            try {
                decode(bytes)
            } catch (e: Exception) {
                rejected = true
            }
            assertTrue(rejected, "${f.name}: accepted malformed input")
        }
    }

    @Test
    fun testEncodesLikeCReference() {
        for (c in parseCases()) {
            // Cases marked encode=no describe values this binding cannot
            // reproduce exactly (float32, or doubles some languages read back
            // as integers).
            if (!c.encodeExpected) continue
            assertContentEquals(
                fixtureFile(c.name).readBytes(),
                encode(caseData(c)),
                "case ${c.name}"
            )
        }
    }
}
